doc/metadata.md

```diff
@@ -72,22 +69,80 @@
         chunk chunks[];
     }
 
+    // this must be a fixed size so we can binary search over it easily
     struct inode {
         u64 ino;
         inode_type type;
+        u32 uid;
+        u32 gid;
+        u16 mode;
+        u64 mtime, atime, ctime; /* seems like we should require these? */
         union {
-            // directory
+            // fifo
+            struct {
+                /* nothing additional about fifos */
+            },
+
+            // chr
+            struct {
+                dev_t major;
+                dev_t minor;
+            },
+
+            // dir
             struct {
-                u64 offset;
+                u64 dir_offset;
             },
 
-            // file
+            // blk; do we even want these? seems like maybe not since they're
+            // system specific.
+            struct {
+                dev_t major;
+                dev_t minor;
+            },
+
+            // reg
             struct {
                 u64 file_size; /* total file size */
-                u64 offest;
+                u64 file_offest;
```

Comment:
Yeah, I'm imagining this would be the size after decompression (i.e. when you're telling the seekable compression algorithm "go to offset X in the payload", this is X).

I think in the case of just an offset, it would be in the current payload. Otherwise, we could make this a metadata_ref type object, but it makes sense to me to just store them inline.

Reply:
yeah, that makes sense.  In this case don't we need to store the length in the compressed stream as well ("go to offset X in the payload and use Y bytes")?

Reply:
Sorry for the late reply, got attacked by other stuff :). Anyway, I'm assuming this offset is in the decompressed stream (perhaps I should make a note of that somewhere, derp); this is the part where I'm assuming this is possible without specifying how it is done: https://github.com/anuvu/puzzlefs/blob/master/doc/metadata.md#reading-compressed-data-at-an-offset

I think initially it may be easiest to use the zstd seekable contrib module, but perhaps puzzlefs should specify its own method for doing this in case we want to switch compression formats.